Updates to draft to fix issues raised (#88)
hannestschofenig/tschofenig-ids: tls-key-update/draft-tschofenig-tls-extended-key-update.md

This specification defines a new, extended key update message supporting perfect
forward secrecy. It does so by utilizing a Diffie-Hellman exchange using one of the
groups negotiated during the initial exchange. The support for this extension is

> signaled using the TLS flags extension mechanism. The frequent rerunning of extended key

Reviewer: ```suggestion
signaled using the TLS flags extension mechanism. The frequent re-running of extended key

```

one direction only. However, the ExtendedKeyUpdate message requires
a full-roundtrip due to the nature of the Diffie-Hellman exchange.

> The KeyShare Entry in the ExtendedKeyUpdate message MUST be the same

Reviewer: ```suggestion
The KeyShare entry in the ExtendedKeyUpdate message MUST be the same

```

section and then re-deriving the traffic keys, as described in
Section 7.3 of {{I-D.ietf-tls-rfc8446bis}}.

There are three changes to the application_traffic_secret computation
described in {{I-D.ietf-tls-rfc8446bis}}, namely

- The application_traffic_secret_N is not used as an secret as it
may be already exfiltrated by the attacker.
- the label is adjusted to distinguish it from the classic KeyUpdate
message, and
- the Diffie-Hellman derived shared secret, as 'dh-secret', is used
as input to the HKDF-Expand-Label() function to produce the value sk.
sk is subsequently included as a secret value in the computation of
the application_traffic_secret_N+1, making the next generation
traffic key of the application traffic secret dependent on the
DH-derived value.

The next-generation application_traffic_secret is computed as follows:

~~~
sk = HKDF-Extract(0, dh-secret)

application_traffic_secret_N+1 =
    Derive-Secret(sk,"traffic upd 2",
                  application_traffic_secret_N)
~~~

> The next generation of traffic key is computed using HKDF as defined in {{RFC5869}} and

Reviewer: ```suggestion
The next generation of traffic keys is computed using the HKDF, as defined in {{RFC5869}}, and

```

section and then re-deriving the traffic keys, as described in
Section 7.3 of {{I-D.ietf-tls-rfc8446bis}}.

There are three changes to the application_traffic_secret computation
described in {{I-D.ietf-tls-rfc8446bis}}, namely

- The application_traffic_secret_N is not used as an secret as it
may be already exfiltrated by the attacker.
- the label is adjusted to distinguish it from the classic KeyUpdate
message, and
- the Diffie-Hellman derived shared secret, as 'dh-secret', is used
as input to the HKDF-Expand-Label() function to produce the value sk.
sk is subsequently included as a secret value in the computation of
the application_traffic_secret_N+1, making the next generation
traffic key of the application traffic secret dependent on the
DH-derived value.

The next-generation application_traffic_secret is computed as follows:

~~~
sk = HKDF-Extract(0, dh-secret)

application_traffic_secret_N+1 =
    Derive-Secret(sk,"traffic upd 2",
                  application_traffic_secret_N)
~~~

The next generation of traffic keys is computed using the HKDF, as defined in {{RFC5869}}, and

> its two components, HKDF-Extract and HKDF-Expand as recommended in Appendix

Reviewer: ```suggestion
its two components, HKDF-Extract and HKDF-Expand, as recommended in Appendix

```

- The application_traffic_secret_N is not used as an secret as it
may be already exfiltrated by the attacker.
- the label is adjusted to distinguish it from the classic KeyUpdate
message, and
- the Diffie-Hellman derived shared secret, as 'dh-secret', is used
as input to the HKDF-Expand-Label() function to produce the value sk.
sk is subsequently included as a secret value in the computation of
the application_traffic_secret_N+1, making the next generation
traffic key of the application traffic secret dependent on the
DH-derived value.

The next-generation application_traffic_secret is computed as follows:

~~~
sk = HKDF-Extract(0, dh-secret)

application_traffic_secret_N+1 =
    Derive-Secret(sk,"traffic upd 2",
                  application_traffic_secret_N)
~~~

The next generation of traffic keys is computed using the HKDF, as defined in {{RFC5869}}, and

its two components, HKDF-Extract and HKDF-Expand, as recommended in Appendix

F.1.1 of {{I-D.ietf-tls-rfc8446bis}}.

Once client_/server_application_traffic_secret_N+1 and its associated
traffic keys have been computed, implementations SHOULD delete
client_/server_application_traffic_secret_N and its associated
traffic keys.

> If Hybrid key exchange is used {{I-D.ietf-tls-hybrid-design}}, the two shared

Reviewer: ```suggestion
If a hybrid key exchange, based on {{I-D.ietf-tls-hybrid-design}}, is used then the two shared

```

> secrets concatenated together (concatenated_shared_secret) is input to

Reviewer: ```suggestion
secrets concatenated together serve as input to

```
